README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

The purpose of this Flutter package is to draw a full-width (embed in Container to resize) representation of a semi-circle assembly...
The number of dots is up to you, by default 577 as for the French National Assembly.

| ![Image](https://github.com/1278real/hemicycle/blob/d23811596e9ee36c5a728390da145ac60a14273c/assets/groupes.png) | ![Image](https://github.com/1278real/hemicycle/blob/177de3d0ba7e0a8d5e76e0ec73f112b5ab44ee9c/assets/legislatives.png) |
| :------------: | :------------: |
| **DrawHemicycle** groupes parlementaires France 2022 | **DrawHemicycle** législatives France 2022 |

## Features

You can change the number of seats, the arc of the circle (by default, 170°) and, with the specific type of inputs, you can display individual voters or sectors for group appearance...

## Usage

Create a ```List<IndividualVotes>``` containing every individual vote you want to draw.

```dart
    List<IndividualVotes> votesTest = [
        IndividualVotes(33, voteResult: true, groupPairing: "AAA"),
        IndividualVotes(34, voteResult: true, groupPairing: "AAA"),
        IndividualVotes(35, voteResult: false, groupPairing: "AAA"),
        IndividualVotes(36, voteResult: true, groupPairing: "AAA"),
        IndividualVotes(37, voteResult: false, groupPairing: "AAA"),
        IndividualVotes(88, voteResult: true, groupPairing: "MMM"),
        IndividualVotes(89, voteResult: false, groupPairing: "MMM"),
        IndividualVotes(90, voteResult: false, groupPairing: "MMM"),
        IndividualVotes(122, voteResult: false, groupPairing: "ZZZ"),
        IndividualVotes(123, voteResult: false, groupPairing: "ZZZ"),
        IndividualVotes(124, voteResult: true, groupPairing: "ZZZ"),
        IndividualVotes(126, voteResult: true, groupPairing: "ZZZ"),
    ];

    DrawHemicycle(200,
        nbRows: 8, individualVotes: votesTest, withLegend: true);
```

OR

Create a ```List<GroupSectors>``` containing every sectors you want to draw.

```dart
    List<GroupSectors> hemicycleTest = [
      GroupSectors(30, customVoteFor, description: "BEFORE"),
      GroupSectors(50, customVoteAgainst, description: "NEW"),
      GroupSectors(497, customVoteAbstention, description: "AFTER")
    ];
    
    DrawHemicycle(
        30 + 50 + 497,
        nbRows: ((30 + 50 + 497) / 50).ceil(),
        groupSectors: hemicycleTest,
        withLegend: true,
        withTitle: true,
        title: "TEST",
    );
```

Then use ```DrawHemicycle``` to get the semi-circle assembly representation. 


| ![Image](https://github.com/1278real/hemicycle/blob/55196e4a7ade0f60c25dbbf5b3a8e7e5179374d9/assets/test_groups.png) | ![Image](https://github.com/1278real/hemicycle/blob/55196e4a7ade0f60c25dbbf5b3a8e7e5179374d9/assets/test_votes.png) |
| :------------: | :------------: |
| **Example** group sectors | **Example** individual votes |

## Additional information

Further infos soon ;-)

CHANGELOG.md
## 0.4.1

* Added second arc for SuperGroup surrounding on Individual Votes view...

## 0.3.2

* Updating informations and comments for pub.dev scoring...

## 0.3.0

* Enable the highlighting for discordant voters in a Group in Individual Votes mode
(i.e. for Vote-For-only votes, the non-voters in a group are interesting to highlight)

## 0.2.2

* Rebuild of the surrounding arc in Individual Votes mode

## 0.1.1

* Graphics enhancements

## 0.1.0

* Added Group surrounding on Individual Votes view

## 0.0.33

* Changed Colors name from custom*** to hemicycle***
(i.e. customVoteFor (green) is now hemicycleVoteFor)

## 0.0.30

* Updating informations for pub.dev scoring...
* Updating README.md and description...

example/hemicycle_example.dart
// Copyright 2022 1•2•7•8 réalisation(s). All rights reserved.
// Use of this source code is governed by a GNU-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';

import 'package:hemicycle/hemicycle.dart';
import 'package:hemicycle/attic/colors.dart';

void main() => runApp(HemicycleExample());

class HemicycleExample extends StatefulWidget {
  @override
  _HemicycleExampleState createState() => _HemicycleExampleState();
}

class _HemicycleExampleState extends State<HemicycleExample> {
  int numberTest = 0;
  int resteTest = 0;

  bool datasUpdated = false;

  List<GroupSectors> hemicycleTest = [];

  List<IndividualVotes> votesTest = [];

  @override
  void initState() {
    numberTest = 1;
    resteTest = 577 - numberTest - 1;

    hemicycleTest = [
      GroupSectors(numberTest, hemicyleVoteFor, description: "BEFORE"),
      GroupSectors(1, hemicyleVoteAgainst, description: "NEW"),
      GroupSectors(resteTest, hemicyleVoteAbstention, description: "AFTER")
    ];

    votesTest = [
      IndividualVotes(33, voteResult: true, groupPairing: "AAA"),
      IndividualVotes(34, voteResult: true, groupPairing: "AAA"),
      IndividualVotes(35, voteResult: false, groupPairing: "AAA"),
      IndividualVotes(36, voteResult: true, groupPairing: "AAA"),
      IndividualVotes(37, voteResult: false, groupPairing: "AAA"),
      IndividualVotes(88, voteResult: true, groupPairing: "MMM"),
      IndividualVotes(89, voteResult: false, groupPairing: "MMM"),
      IndividualVotes(90, voteResult: false, groupPairing: "MMM"),
      IndividualVotes(122, voteResult: false, groupPairing: "ZZZ"),
      IndividualVotes(123, voteResult: false, groupPairing: "ZZZ"),
      IndividualVotes(124, voteResult: true, groupPairing: "ZZZ"),
      IndividualVotes(126, voteResult: true, groupPairing: "ZZZ"),
    ];

    updateAndRefresh();
    super.initState();
  }

  void updateAndRefresh() async {
    Future.delayed(Duration(milliseconds: 100), (() {
      setState(() {
        datasUpdated = true;
      });
    }));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(children: <Widget>[
        if (datasUpdated)
          DrawHemicycle(200,
              nbRows: 8, individualVotes: votesTest, withLegend: true),
        if (datasUpdated)
          DrawHemicycle(
            resteTest + numberTest + 1,
            nbRows: ((resteTest + numberTest + 1) / 50).ceil(),
            groupSectors: hemicycleTest,
            withLegend: true,
            withTitle: true,
            title: "TEST",
          ),
        TextButton(
            onPressed: () {
              setState(() {
                numberTest += 1;
                datasUpdated = false;
              });
              updateAndRefresh();
            },
            child: Text(("PLUS UN... (" + numberTest.toString() + ")"))),
      ]),
    );
  }
}
